### PR TITLE
feat(driver-adapters): enable Wasm on `driver-adapters`, with further deduplications required later

### PR DESCRIPTION
**Pull request by jkomyno**
This PR contributes to https://github.com/prisma/team-orm/issues/548, and depends on https://github.com/prisma/prisma-engines/pull/4455.

Idea: add Wasm support to `driver-adapters`, which currently relies on `napi.rs` only, by splitting the wasm-specific logic into the `wasm` module, and the `napi.rs`-specific logic into the `napi` module.

These modules are mutually exclusive and provide *almost* the same public API.
The `wasm` module is compiled when the target architecture is `wasm32-*`, and `napi` is used otherwise.

See [Slack thread](https://prisma-company.slack.com/archives/C04TW4A2H8C/p1700065803471809) for context about it.

---

For what is worth, I think this PR is good enough for getting the vertical slice asap, but not for integrating into `main`, as I  pointed our in a few comments.

### CODE REVIEW

File: query-engine/driver-adapters/src/wasm/queryable.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+/// connector)
+///
+/// Transforming a `JSResultSet` (what client connectors implemented in javascript provide)
+/// into a `quaint::connector::result_set::ResultSet`. A quaint `ResultSet` is basically a vector
+/// of `quaint::Value` but said type is a tagged enum, with non-unit variants that cannot be converted to javascript as is.
+#[wasm_bindgen(getter_with_clone)]
+#[derive(DuckType, Default)]
+pub(crate) struct JsBaseQueryable {
+    pub(crate) proxy: CommonProxy,
+    pub flavour: Flavour,
```
Comment (jkomyno) [inline query-engine/driver-adapters/src/wasm/queryable.rs:36]: Possible improvement: don't depend on `Flavour` for this struct, as it's exposed to JS via `#[wasm_bindgen]`.

File: query-engine/driver-adapters/src/wasm/transaction.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    fn requires_isolation_first(&self) -> bool {
+        self.inner.requires_isolation_first()
+    }
+}
+
+// Assume the proxy object will not be sent to service workers, we can unsafe impl Send + Sync.
+unsafe impl Send for JsTransaction {}
+unsafe impl Sync for JsTransaction {}
```
Comment (jkomyno) [inline query-engine/driver-adapters/src/wasm/transaction.rs:132]: Background info: `Queryable` requires `Send` base structs due to `#[async_trait]`.
At some point, we should evaluate replacing it into `#[async_trait(?Send)]`, if it makes sense.
Alternatively, we could do so for `wasm32-*` target architecture only, preserving the `Send` requirement for Rust drivers. This is just some food for thought, nothing applicable right now.

File: query-engine/query-engine-wasm/src/engine.rs [hunk 3]
```diff
@@ -125,10 +126,12 @@ impl QueryEngine {
...
         let mut schema = psl::validate(datamodel.into());
         let config = &mut schema.configuration;
+        let preview_features = config.preview_features();
 
         if let Some(adapter) = maybe_adapter {
-            let js_queryable =
-                proxy::from_wasm(adapter).map_err(|e| ApiError::configuration(e.as_string().unwrap_or_default()))?;
+            let js_queryable = driver_adapters::from_wasm(adapter);
+
+            sql_connector::activate_driver_adapter(Arc::new(js_queryable));
```
Comment (jkomyno) [inline query-engine/query-engine-wasm/src/engine.rs:134]: This is a direct adaptation of what we're doing in `query-engine-node-api/src/engine.rs`

File: query-engine/driver-adapters/Cargo.toml [hunk 4]
```diff
@@ -22,9 +21,21 @@ bigdecimal = "0.3.0"
...
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+quaint = { path = "../../quaint" }
+js-sys.workspace = true
+serde-wasm-bindgen.workspace = true
+wasm-bindgen.workspace = true
+wasm-bindgen-futures.workspace = true
+tsify.workspace = true
+ducktor = "0.1.0"
+pin-project = "1"
```
Comment (Weakky) [inline query-engine/driver-adapters/Cargo.toml:41]: I believe you should be able to use quaint's workspace version in both cases here. [See doc here](https://doc.rust-lang.org/cargo/reference/resolver.html#feature-resolver-version-2)

image

File: query-engine/driver-adapters/Cargo.toml [hunk 5]
```diff
@@ -22,9 +21,21 @@ bigdecimal = "0.3.0"
...
+quaint.workspace = true
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+quaint = { path = "../../quaint" }
+js-sys.workspace = true
+serde-wasm-bindgen.workspace = true
+wasm-bindgen.workspace = true
+wasm-bindgen-futures.workspace = true
+tsify.workspace = true
+ducktor = "0.1.0"
```
Comment (Weakky) [inline query-engine/driver-adapters/Cargo.toml:40]: This dependency, `ducktor`, looks scary tbh. Completely unknown, a few months old with barely any tests. How much work does it remove from our plate?

File: query-engine/query-engine-wasm/src/wasm/engine.rs [hunk 6]
```diff
@@ -125,10 +126,12 @@ impl QueryEngine {
 
         let mut schema = psl::validate(datamodel.into());
         let config = &mut schema.configuration;
+        let preview_features = config.preview_features();
```
Comment (Weakky) [inline query-engine/query-engine-wasm/src/wasm/engine.rs:129]: looks unused?

**Review by Weakky — CHANGES_REQUESTED** [review]
Too much code duplication between `driver-adapters:napi` and `driver-adapters::wasm`. Some work needs to be done to have a shared implementation.

File: query-engine/driver-adapters/Cargo.toml (hunk 4 above)
Comment (jkomyno) [reply query-engine/driver-adapters/Cargo.toml:41]: Thanks, I didn't know this about `resolver = "2"`.

File: query-engine/driver-adapters/Cargo.toml (hunk 5 above)
Comment (jkomyno) [reply query-engine/driver-adapters/Cargo.toml]: `ducktor` was one of the shortcuts taken (I already had some code using it laying around).

Removed in https://github.com/prisma/prisma-engines/pull/4456/commits/d6d09d6f47999af237eac47771d69db68d00e2ed.

Comment (jkomyno) [reply query-engine/driver-adapters/Cargo.toml]: Looks like this change broke something:

```
Error: `unwrap_throw` failed
    at __wbindgen_throw (file:///Users/jkomyno/work/prisma/prisma-engines/query-engine/query-engine-wasm/pkg/query_engine_bg.js:977:11)
    at wasm://wasm/01063c0e:wasm-function[5374]:0x352371
    at wasm://wasm/01063c0e:wasm-function[3153]:0x32513b
    at wasm://wasm/01063c0e:wasm-function[93]:0x3902b
```

I'll take another look tomorrow.

**Comment by jkomyno** [comment]
This is currently broken, see https://github.com/prisma/prisma-engines/pull/4456#discussion_r1401417270.

**Comment by jkomyno** [comment]
As per [Slack](https://prisma-company.slack.com/archives/C04TW4A2H8C/p1700649363732899?thread_ts=1700451627.187479&cid=C04TW4A2H8C), I'll be merging this to the vertical slice PR, #4466.

**Pull request closed without merging**